src/evict/evict_page.c

```diff
@@ -675,12 +675,41 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
             LF_SET(WT_REC_SCRUB);
     }
 
-    /* Reconcile the page. */
-    ret = __wt_reconcile(session, ref, NULL, flags);
+    /* Acquire a snapshot if coming through eviction thread route. */
+    if (F_ISSET(session, WT_SESSION_INTERNAL) && !WT_SESSION_BTREE_SYNC(session) &&
+      !F_ISSET(conn,
+        WT_CONN_CLOSING | WT_CONN_RECOVERING | WT_CONN_IN_MEMORY | WT_CONN_CLOSING_TIMESTAMP) &&
+      !WT_IS_HS(S2BT(session)) && !F_ISSET(session->txn, WT_TXN_RUNNING | WT_TXN_HAS_SNAPSHOT)) {
+        /*
+         * Eviction threads do not need to pin anything in the cache. We have a exclusive lock for
+         * the page being evicted so we are sure that the page will always be there while it is
+         * being processed. Therefore, we use snapshot API that doesn't publish shared IDs to
+         * outside world.
+         */
+        __wt_txn_bump_snapshot(session);
+        snapshot_acquired = true;
+    } else if (!WT_SESSION_BTREE_SYNC(session))
+        LF_SET(WT_REC_VISIBLE_ALL);
+
+    WT_ASSERT(session, LF_ISSET(WT_REC_VISIBLE_ALL) || F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT));
+
+    /*
+     * Reconcile the page. If we acquired a snapshot for eviction, used forced isolation as opening
+     * and closing history store cursor for deleting history store record (as part of
+     * reconciliation) can lead to release of snapshot.
```

Comment:
To make sure I understand, could you do a `++txn->forced_iso` and decrement it when doing the release instead?

i.e: does it matter what isolation level is set, or just that the snapshot isn't released earlier than is OK?

Reply:
All eviction thread sessions have `WT_ISO_READ_COMMITTED` isolation level set by default so we can skip forcing the isolation level unless we don't want to rely on the defaults. Otherwise, we can achieve our required behaviour by incrementing/decrementing `txn->forced_iso`.

Reply:
@agorrod - I have tried setting isolation level at the time of creating eviction threads in 1df4a42.